File: apollo/src/main/java/io/muun/apollo/domain/utils/FeeCalculator.java
package io.muun.apollo.domain.utils;


import io.muun.apollo.domain.errors.AmountTooSmallError;
import io.muun.apollo.domain.errors.InsufficientFundsError;
import io.muun.common.model.SizeForAmount;
import io.muun.common.utils.BitcoinUtils;

import java.util.List;

import javax.validation.constraints.NotNull;

public class FeeCalculator {

    @NotNull
    private final long satoshisPerByte;

    @NotNull
    private final List<SizeForAmount> sizeProgression;

    /**
     * Create a FeeCalculator for a FeeWindow, given an array of Transaction size estimations.
     */
    public FeeCalculator(long satoshisPerByte, List<SizeForAmount> sizeProgression) {
        this.satoshisPerByte = satoshisPerByte;
        this.sizeProgression = sizeProgression;
    }

    /**
     * Calculate the maximum spendable amount (in satoshis) for a total balance, subtracting the
     * fee it would cost to actually send it.
     */
    public long getMaxSpendableAmount() {
        if (sizeProgression.size() == 0) {
            return 0;
        }

        long peak = 0;
        for (SizeForAmount sizeForAmount : sizeProgression) {

            final long fee = sizeForAmount.sizeInBytes * satoshisPerByte;
            final long candidatePeak = sizeForAmount.amountInSatoshis - fee;

            // A new candidatePeak may not be higher than the current peak. This happens because
            // there are utxos that are worth less than the fee that it would cost to spend them.
            // And this is regardless of the order of the utxos in the sizeProgression.
            if (peak < candidatePeak) {
                peak = candidatePeak;
            }
        }

        return peak;
    }

    /**
     * Calculate the required fee to send a given amount, using the FeeWindow and optional
     * transaction size estimations provided on construction.
     */
    public long getFeeForAmount(long amountInSatoshis) {
        if (amountInSatoshis < BitcoinUtils.DUST_IN_SATOSHIS) {
            throw new AmountTooSmallError();
        }

        for (SizeForAmount sizeForAmount: sizeProgression) {
            if (sizeForAmount.amountInSatoshis < amountInSatoshis) {
                // With this entry, we don't have enough funds in UTXOs to cover the operation
                // amount. Next!
                continue;
            }

            final long fee = satoshisPerByte * sizeForAmount.sizeInBytes;

            if (sizeForAmount.amountInSatoshis < amountInSatoshis + fee) {
                // With this entry, we can cover the original amount, but not the amount plus the
                // fee we're about to recommend. Next!
                continue;
            }

            return fee;
        }

        throw new InsufficientFundsError(); // oh noes!
    }
}

File: apollo/src/main/java/io/muun/apollo/domain/utils/FeeCalculator.kt
package io.muun.apollo.domain.utils


import io.muun.apollo.domain.errors.InsufficientFundsError
import io.muun.apollo.domain.model.NextTransactionSize
import kotlin.math.ceil

/**
 * Calculate fees using a FeeWindow, given an array of Transaction size estimations.
 */
class FeeCalculator(val satoshisPerByte: Double,
                    val nextTransactionSize: NextTransactionSize) {

    /**
     * Return the FeeCalculation for a requested amount, using UTXO_BALANCE.
     */
    fun calculateForCollect(amountInSatoshis: Long, takeFeeFromAmount: Boolean = false): Long {
        return calculate(amountInSatoshis, takeFeeFromAmount, nextTransactionSize.utxoBalance)
    }

    /**
     * Return the FeeCalculation for a requested amount, using USER_BALANCE.
     */
    fun calculate(amountInSatoshis: Long, takeFeeFromAmount: Boolean = false): Long {
        return calculate(amountInSatoshis, takeFeeFromAmount, nextTransactionSize.userBalance)
    }

    /**
     * Return the FeeCalculation for a requested amount, using specified balance
     * (USER_BALANCE VS UTXO_BALANCE).
     */
    private fun calculate(amountInSatoshis: Long, takeFeeFromAmount: Boolean, balance: Long): Long {
        if (amountInSatoshis == 0L) {
            return 0 // a special case, for consistency
        }

        if (amountInSatoshis > balance) {
            throw InsufficientFundsError() // we cannot return a meaningful result
        }

        return if (takeFeeFromAmount) {
            getFeeTakenFromAmount(amountInSatoshis)

        } else {
            getFeeTakenFromRemainingBalance(amountInSatoshis)
        }
    }

    private fun getFeeTakenFromAmount(amountInSatoshis: Long): Long {
        var feeInSatoshis = 0L

        for (sizeForAmount in nextTransactionSize.sizeProgression) {
            feeInSatoshis = applyRate(sizeForAmount.sizeInBytes, satoshisPerByte)

            if (sizeForAmount.amountInSatoshis >= amountInSatoshis) {
                break // no more UTXOs needed
            }
        }

        return feeInSatoshis
    }

    private fun getFeeTakenFromRemainingBalance(amountInSatoshis: Long): Long {
        var feeInSatoshis: Long = 0

        for (sizeForAmount in nextTransactionSize.sizeProgression) {
            feeInSatoshis = applyRate(sizeForAmount.sizeInBytes, satoshisPerByte)

            if (sizeForAmount.amountInSatoshis >= amountInSatoshis + feeInSatoshis) {
                break // no more UTXOs needed
            }
        }

        return feeInSatoshis
    }

    private fun applyRate(sizeInBytes: Int, satoshisPerByte: Double) =
        ceil(satoshisPerByte * sizeInBytes).toLong()
}
